packages/tpetra/core/src/Tpetra_CrsSingletonFilter_LinearProblem_def.hpp

```diff
@@ -0,0 +1,1768 @@
+    construct() {
+  const char tfecfFuncName[] = "construct: ";
+
+  TEUCHOS_TEST_FOR_EXCEPTION_CLASS_FUNC(this->origObj_ == Teuchos::null, std::runtime_error,
+                                        "Linear problem (orgiObj_) have not been set.  Call analyze() first.");
+
+  ConstructReducedProblem(this->origObj_);
+
+  this->newObj_ = ReducedProblem();
+
+  if (verbose_ && SingletonsDetected() && FullMatrix()->getComm()->getRank() == 0) {
+    std::cout << "\nConstructedSingleton Problem:\n";
+    std::cout << "---------------------------\n";
+    std::cout << "RatioOfDimensions:   " << RatioOfDimensions() << std::endl;
+    std::cout << "RatioOfNonzeros:     " << RatioOfNonzeros() << std::endl;
+    std::cout << "---------------------------\n\n";
+  }
+
+  return this->newObj_;
+}
+
+//==============================================================================
+
+template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+void CrsSingletonFilter_LinearProblem<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
+    fwd() {
+  UpdateReducedProblem(FullProblem());
+}
+
+//==============================================================================
+
+template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+void CrsSingletonFilter_LinearProblem<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
+    rvs() {
+  ComputeFullSolution();
+}
+
+//==============================================================================
+
+template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+void CrsSingletonFilter_LinearProblem<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
+    Analyze(const Teuchos::RCP<row_matrix_type>& fullMatrix) {
+  const char tfecfFuncName[] = "Analyze: ";
+
+  TEUCHOS_TEST_FOR_EXCEPTION_CLASS_FUNC(AnalysisDone_, std::runtime_error,
+                                        "Analyze() already done once.  Cannot do it again.");
+
+  TEUCHOS_TEST_FOR_EXCEPTION_CLASS_FUNC(fullMatrix == Teuchos::null,
+                                        std::runtime_error, "Input matrix is Teuchos::null.");
+
+  TEUCHOS_TEST_FOR_EXCEPTION_CLASS_FUNC(fullMatrix->getGlobalNumRows() == 0,
+                                        std::runtime_error, "Full matrix has zero dimension.");
+
+  TEUCHOS_TEST_FOR_EXCEPTION_CLASS_FUNC(fullMatrix->getGlobalNumEntries() == 0,
+                                        std::runtime_error, "Full matrix has no nonzero terms.");
+
+  FullMatrix_ = fullMatrix;
+
+  // First check for columns with single entries and find columns with singleton rows
+  vector_type_int ColProfiles(FullMatrixColMap());
+  ColProfiles.putScalar(0);
+  vector_type_int ColHasRowWithSingleton(FullMatrixColMap());
+  ColHasRowWithSingleton.putScalar(0);
+
+  // localRowIDofSingletonCol[j] will contain the local row ID associated with the jth column,
+  // if the jth col has a single entry
+  vector_type_LO localRowIDofSingletonCol(FullMatrixColMap());
+  localRowIDofSingletonCol.putScalar(-1);
```

Comment:
Change `-1` to `Teuchos::OrdinalTraits<LO>::invalid()`. Not sure what `-1` does if someone decided to use `LO=unsigned int`.

Reply:
Done.